documentation/directory-structure.md

```diff
@@ -30,13 +34,66 @@ You may be also interested in following:
 > The structure described in this article is strictly enforced. There exist some exceptions for historical reasons.
 > These exceptions are not allowed going forward.
 
+## Key concepts
+
+The directory structure is a reflection of few key concepts.
+
+A `service` is a set of operation endpoints (typically HTTP REST API endpoints) that **version uniformly**.
+
+A `service group` is a set of services that share common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
+
+In case of ARM, a `service` consists of multiple `resource types`.
+
+### Uniform versioning
+
+A `service` **must version uniformly**. This means that the service,
+all its dependencies, and all artifacts generated from the service,
+**must** have the same version.
+
+In practice, this means, for an example service with version `2024-03-25`:
+
+- All the OpenAPI specification `.json` files the service is composed of must have the same version of `2024-03-25`.
+- All the shared specifications (i.e. `common-types`) the service depends on must have the same version of `2024-03-25`.
```

Comment:
Common types typically do not use date-based versioning. But we can say that the versions should all match.
```suggestion
- All the shared specifications (i.e. `common-types`) the service depends on must have the same version.
```